runtime/src/bank.rs

```diff
@@ -3669,6 +3671,41 @@ impl Bank {
         Ok(())
     }
 
+    fn get_account_rent_states(&self, accounts: &[TransactionAccountRefCell]) -> Vec<RentState> {
+        accounts
+            .iter()
+            .map(|account_refcell| {
+                RentState::from_account(account_refcell, &self.rent_collector.rent)
+            })
+            .collect()
+    }
+
+    fn check_rent_state_changes(
+        &self,
+        process_result: &mut Result<()>,
+        pre_rent_states: &[RentState],
+        post_rent_states: &[RentState],
+        accounts: &[TransactionAccountRefCell],
+    ) {
+        match process_result {
+            Ok(_) | Err(TransactionError::InstructionError(_, _)) => {
+                for (i, (pre_rent_state, post_rent_state)) in
+                    pre_rent_states.iter().zip(post_rent_states).enumerate()
+                {
+                    if !post_rent_state.transition_allowed_from(pre_rent_state) {
+                        debug!(
+                            "Account {:?} not rent exempt, lamports {:?}",
+                            accounts[i].0,
+                            accounts[i].1.borrow().lamports()
+                        );
+                        *process_result = Err(TransactionError::InvalidRentPayingAccount)
```

Comment:
I don't think we want to return a `TransactionError` after processing a transaction. That means it's free to process potentially expensive transactions since fees aren't deducted for tx errors.

Reply:
Yeah, I was pondering adding this TransactionError as a special case in `filter_program_errors_and_collect_fee()`. But perhaps a better approach would be to pass the rent information into the message processor and evaluate on a per-instruction basis with the other pre/post processing checks. Thoughts?

Reply:
I think it's fine to use non-rent-exempt accounts between instructions for passing data around. Would your change prevent that?